src/bootstrap/src/utils/build_stamp.rs

```diff
@@ -0,0 +1,73 @@
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+};
+
+#[derive(Clone)]
+pub struct BuildStamp {
+    path: PathBuf,
+    stamp: String,
+}
+
+impl From<BuildStamp> for PathBuf {
+    fn from(value: BuildStamp) -> Self {
+        value.path
+    }
+}
+
+impl AsRef<Path> for BuildStamp {
+    fn as_ref(&self) -> &Path {
+        &self.path
+    }
+}
+
+impl BuildStamp {
+    pub fn new(dir: &Path) -> Self {
+        Self { path: dir.join(".stamp"), stamp: String::new() }
+    }
+
+    pub fn with_stamp(mut self, stamp: String) -> Self {
+        self.stamp = stamp;
+        self
+    }
```

Comment:
Question: this allows a `BuildStamp` instance to call `with_stamp` multiple times. Is that intentional? I.e.

```rs
let bs = BuildStamp::new(&meow_dir).with_stamp("a").with_stamp("b")
```

which I'm not sure the intended behavior -- is this supposed to produce two stamp files `a` and `b`?

Maybe forbid setting `stamp` more than once?

Reply:
[Duplicated](https://github.com/rust-lang/rust/pull/135281#discussion_r1912325528)